File: apps/web/src/lib/urn.ts
// 任務身分 URN(scheme 同 indexer api/urn.py):由 provenance 現組、不落庫(spec §2 三分:
// 身分存 provenance、顯示用位置碼、鍵用 URN 現組)。
// 自訂任務(無 provenance)回 ""(呼叫端據此停用 catalog query)。
export function taskUrn(p?: { ocs_code?: string; task_code?: string }): string {
  return p?.ocs_code && p?.task_code ? `ocs:${p.ocs_code}:T:${p.task_code}` : "";
}

// 任務的全部身分 URN:合併列選入的任務 _refs 多筆;舊資料/單來源 fallback provenance。
export function taskUrns(task: {
  provenance?: { ocs_code?: string; task_code?: string };
  _refs?: { ocs_code: string; task_code?: string }[];
}): string[] {
  const refs = task._refs?.length ? task._refs : task.provenance ? [task.provenance] : [];
  const out: string[] = [];
  for (const r of refs) {
    const u = taskUrn({ ocs_code: r.ocs_code, task_code: r.task_code });
    if (u && !out.includes(u)) out.push(u);
  }
  return out;
}

File: apps/ocs-indexer/src/jd_ocs_indexer/api/urn.py
"""Stable citable identifiers (URN) for OCS entities. Composed at query time
from payload fields; never stored. Scheme: ocs:{ocs_code}[:{TYPE}:{code}]."""
from __future__ import annotations


def occupation_urn(ocs_code: str) -> str:
    return f"ocs:{ocs_code}"


def unit_urn(ocs_code: str, ocu_code: str) -> str:
    return f"ocs:{ocs_code}:U:{ocu_code}"


def task_urn(ocs_code: str, task_code: str) -> str:
    return f"ocs:{ocs_code}:T:{task_code}"


def item_urn(ocs_code: str, type_: str, code: str) -> str:
    return f"ocs:{ocs_code}:{type_}:{code}"
